clippy_lints/src/rest_when_destructuring_struct.rs

```diff
@@ -0,0 +1,103 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::is_from_proc_macro;
+use itertools::Itertools;
+use rustc_abi::VariantIdx;
+use rustc_lint::LateLintPass;
+use rustc_middle::ty::{self, Visibility};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Disallows the use of rest patterns when destructuring structs.
+    ///
+    /// ### Why is this bad?
+    /// It might lead to unhandled fields when the struct changes.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// struct S {
+    ///     a: u8,
+    ///     b: u8,
+    ///     c: u8,
+    /// }
+    ///
+    /// let s = S { a: 1, b: 2, c: 3 };
+    ///
+    /// let S { a, b, .. } = s;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// struct S {
+    ///     a: u8,
+    ///     b: u8,
+    ///     c: u8,
+    /// }
+    ///
+    /// let s = S { a: 1, b: 2, c: 3 };
+    ///
+    /// let S { a, b, c: _ } = s;
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub REST_WHEN_DESTRUCTURING_STRUCT,
+    nursery,
+    "rest (..) in destructuring expression"
+}
+declare_lint_pass!(RestWhenDestructuringStruct => [REST_WHEN_DESTRUCTURING_STRUCT]);
+
+impl<'tcx> LateLintPass<'tcx> for RestWhenDestructuringStruct {
+    fn check_pat(&mut self, cx: &rustc_lint::LateContext<'tcx>, pat: &'tcx rustc_hir::Pat<'tcx>) {
+        if let rustc_hir::PatKind::Struct(path, fields, Some(dotdot)) = pat.kind
+            && !pat.span.in_external_macro(cx.tcx.sess.source_map())
+            && !is_from_proc_macro(cx, pat)
+            && let qty = cx.typeck_results().qpath_res(&path, pat.hir_id)
+            && let ty = cx.typeck_results().pat_ty(pat)
+            && let ty::Adt(a, _) = ty.kind()
+        {
+            let vid = qty
+                .opt_def_id()
+                .map_or(VariantIdx::ZERO, |x| a.variant_index_with_id(x));
+
+            let leave_dotdot = a.variants()[vid].field_list_has_applicable_non_exhaustive();
+
+            let mut rest_fields = a.variants()[vid]
+                .fields
+                .iter()
+                .filter(|f| {
+                    if a.did().is_local() {
+                        true
+                    } else {
+                        matches!(f.vis, Visibility::Public)
+                    }
+                })
+                .map(|field| field.ident(cx.tcx))
+                .filter(|pf| !fields.iter().any(|x| x.ident == *pf))
```

Comment:
We can skip recovering the ident here and compare the field name against `.ident.name`